### PR TITLE
Added m_joinpartspam: chmode to ban joinpart spammers.

### PR DESCRIPTION
**Pull request by genius3000**
This adds channel mode +U with a 'cycles:sec' parameter. A cycle is counted as a successful join to the channel. Upon reaching the limit, the next attempted join (prior to the time reset) results in a WildHost ban and denying the join.

I'm using a mask of 'user@host' instead of 'User*' for tracking so a user is still caught after a quit, connect, and join. Not using the nickname in the mask so it's not so easily bypassed.

### CODE REVIEW

**Comment by GLolol** [comment]
Cmode +U conflicts with #64 :(

**Comment by Techman-** [comment]
Network ops still have the ability to pick and choose what modules to use :)

Configurable modechar would be good.

**Comment by genius3000** [comment]
I just had a look through helpop and went with an available letter. Could try find another one.
It is easy to change, just one spot in the code (other than comments).

**Comment by genius3000** [comment]
Updated this with a few code style fixes and changed to chmode V. I know that could conflict with customprefix->Half voice, but I've never seen that used.

File: 2.0/m_joinpartspam.cpp [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+	void cleanup()
+	{
+		/* 10 minutes should be a reasonable wait time */
+		if (ServerInstance->Time() - lastcleanup < 600)
+			return;
+
+		for (std::map<std::string, time_t>::iterator i = resets.begin(); i != resets.end(); i++)
+		{
+			if (ServerInstance->Time() > resets[i->first])
+				this->clear(i->first);
```
Comment (attilamolnar) [inline 2.0/m_joinpartspam.cpp]: This call invalidates the iterator, if the entry needs to be removed ++ the iterator before calling the function

File: 2.0/m_joinpartspam.cpp [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+	time_t lastcleanup;
+
+	struct Tracking
+	{
+		time_t reset;
+		unsigned int counter;
+	};
+	std::map<std::string, Tracking> cycler;
+
+	joinpartspamsettings(int c, int s) : cycles(c), secs(s)
```
Comment (attilamolnar) [inline 2.0/m_joinpartspam.cpp]: These parameters should be `unsigned int`.

File: 2.0/m_joinpartspam.cpp [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+		 */
+		if (cycler.count(mask) != 1)
+			cycler[mask].reset = ServerInstance->Time() + secs;
+		else if (ServerInstance->Time() > cycler[mask].reset || cycler[mask].counter >= cycles)
+		{
+			cycler[mask].counter = 0;
+			cycler[mask].reset = ServerInstance->Time() + secs;
+		}
+
+		++cycler[mask].counter;
```
Comment (attilamolnar) [inline 2.0/m_joinpartspam.cpp]: Each `cycler[mask]` causes a map lookup which is wasteful because you are only interacting with the same element throughout the method. To improve this, do something like:
`Tracking& tracking = cycler[mask];`
then replace `cycler[mask]` with `tracking` in the rest of the method.

File: 2.0/m_joinpartspam.cpp [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+/* $ModAuthor: genius3000 */
+/* $ModAuthorMail: genius3000@g3k.solutions */
+/* $ModDepends: core 2.0 */
+
+class joinpartspamsettings
+{
+ public:
+	unsigned int cycles;
+	unsigned int secs;
+	time_t lastcleanup;
```
Comment (attilamolnar) [inline 2.0/m_joinpartspam.cpp]: This must be given a value before line 88 checks it. As I understand the initial value has to pass the check on that line for the code to work as intended.

File: 2.0/m_joinpartspam.cpp [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+		{
+			if (ServerInstance->Time() > cycler[mask].reset)
+				cycler.erase(mask);
+			else if (cycler[mask].counter >= cycles)
+			{
+				cycler.erase(mask);
+				return true;
+			}
+		}
+		return false;
```
Comment (attilamolnar) [inline 2.0/m_joinpartspam.cpp]: What you can do here to avoid repeated map lookups is to use the `find()` map method like this:

```c++
std::map<std::string, Tracking>::iterator it = cycler.find(mask);
if (it == cycler.end()) // 'mask' is not in the map
    return false;
Tracking& tracking = it->second;

// Rest of the code
if (ServerInstance->Time() > tracking.reset)
 ...
```

File: 2.0/m_joinpartspam.cpp [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+	}
+
+	/* Clear expired entries of non cyclers */
+	void cleanup()
+	{
+		/* 10 minutes should be a reasonable wait time */
+		if (ServerInstance->Time() - lastcleanup < 600)
+			return;
+
+		for (std::map<std::string, Tracking>::iterator i = cycler.begin(); i != cycler.end(); )
```
Comment (attilamolnar) [inline 2.0/m_joinpartspam.cpp]: The same thing applies for `cycler[mask]` and `erase(key)` as above, they must find the element in the map but you already know where it is.
There's existing code showing how to do this more efficiently: https://github.com/inspircd/inspircd/blob/master/src/coremods/core_dns.cpp#L658-L665
That way, no extra strings get created and extra map lookups don't occur.

**Comment by genius3000** [comment]
@attilamolnar how's this looking now?

**Comment by attilamolnar** [comment]
@genius3000 Code-wise looks ok. I see it is storing the ident along with the host. Do you intend it to work this way?

File: 2.0/m_joinpartspam.cpp [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+			const std::string& mask(user->MakeHost());
+			if (jpss->zapme(mask))
+			{
+				std::vector<std::string> parameters;
+				parameters.push_back(chan->name);
+				parameters.push_back("+b");
+				parameters.push_back(user->MakeWildHost());
+				ServerInstance->SendGlobalMode(parameters, ServerInstance->FakeClient);
+
+				user->WriteNumeric(474, "%s %s :Channel JoinPart Spam triggered (limit is %u cycles in %u secs)", user->nick.c_str(), chan->name.c_str(), jpss->cycles, jpss->secs);
```
Comment (attilamolnar) [inline 2.0/m_joinpartspam.cpp]: Can you reword this to say "join/part spam" to look more natutal instead of "JoinPart Spam"?

**Comment by genius3000** [comment]
@attilamolnar yes, I intended the use of ident@host; Figured host may be too broad, and using nickname as well is too specific/bypassed.
I'll change the wording in the three Numerics, that does look better.
Good to squash the commits down to just one then as well?

**Comment by attilamolnar** [comment]
Ok.

> Good to squash the commits down to just one then as well?

Yeah please do

**Comment by genius3000** [comment]
@atillamolnar Alright, squashed down to the one commit. Thank you very much for all the assistance.

**Comment by attilamolnar** [comment]
Thanks for your contribution!
